File: 2019/scala/src/main/scala/Day2.scala
object Day2 {
  val input = scala.io.Source.fromFile("data/data2.txt", "UTF-8").getLines.toVector(0).split(',').map(_.toInt).toVector

  def part1(noun: Int = 12, verb: Int = 2): Int = {
    val mem = input.toArray
    var pc = 0
    mem(1) = noun
    mem(2) = verb
    while (pc < mem.length && mem(pc) != 99) {
      mem(mem(pc + 3)) = mem(pc) match {
        case 1 => mem(mem(pc + 1)) + mem(mem(pc + 2))
        case 2 => mem(mem(pc + 1)) * mem(mem(pc + 2))
        case _ => {println("unknown opcode" + mem(pc)); mem(mem(pc + 3))}
      }
      pc += 4
    }
    mem(0)
  }

  def part2(): Unit = {
    for (i <- 0 to 99; j <- 0 to 99) {
      try {
        if (part1(i,j) == 19690720) println("noun=" + i + " verb=" + j)
      } catch {
        case e: Exception => -1
      }
    }
  }

  def apply() = {
    println("SOLUTION DAY 2")
    println("Running part 1")
    println(part1())
    println("Running part 2")
    part2()
  }
}

File: 2019/python/src/day2.py
import os
import sys
import re

def main():
    input = []
    
    with open(f'{os.path.dirname(os.path.realpath(__file__))}/../input/day2.txt') as f:
        for line in f.readlines():
            input.append(line.replace("\n",""))

    print("Part 1: " + str(part1(input)))
    print("Part 2: " + str(part2(input)))

def part1(input):
    ints = list(map(int,re.findall(r'\d+', input[0])))
    ints[1] = 12
    ints[2] = 2
    pc = 0
    while ints[pc] != 99:
        opcode = ints[pc]
        if opcode == 1:
            ints[ints[pc+3]] = ints[ints[pc+1]] + ints[ints[pc+2]]
        elif opcode == 2:
            ints[ints[pc+3]] = ints[ints[pc+1]] * ints[ints[pc+2]]
        pc += 4
    
    return ints[0]
    
def part2(input):
    for x in range(0,100):
        for y in range(0,100):
            ints = list(map(int,re.findall(r'\d+', input[0])))
            ints[1] = x
            ints[2] = y
            pc = 0
            while ints[pc] != 99:
                opcode = ints[pc]
                if opcode == 1:
                    ints[ints[pc+3]] = ints[ints[pc+1]] + ints[ints[pc+2]]
                elif opcode == 2:
                    ints[ints[pc+3]] = ints[ints[pc+1]] * ints[ints[pc+2]]
                pc += 4
            if ints[0] == 19690720:
                return 100*x + y
    
    return "NO FOUND!"

if __name__ == "__main__":
    sys.exit(main())